events/equeue/equeue_mbed.cpp

```diff
@@ -150,8 +150,8 @@ bool equeue_sema_wait(equeue_sema_t *s, int ms) {
 
     core_util_critical_section_enter();
     while (!*s) {
-        sleep();
         core_util_critical_section_exit();
+        sleep();
```

Comment:
Sorry bud, this still fails.

``` cpp
core_util_critical_section_enter();         // *s is false when entering critical section
// interrupt fires!                         // hah! we're in a critical section, interrupt is pending but does not run
while (!*s) {                               // *s is false, enter while loop to wait
    core_util_critical_section_exit();
    // pending interrupt runs           // oh no! now that we've left the critical section, the pending interrupt runs
                                        // *s is set to true
    sleep();                            // nothing will wake up this sleep call even though *s is true
    core_util_critical_section_enter();
    blablabla
```

Counterintuitively, sleep must support critical sections.

Keep in mind, if we break code inside the OS, we are _definitely_ breaking code outside the OS.